src/coreclr/src/vm/eventpipesession.cpp

```diff
@@ -482,12 +563,32 @@ void EventPipeSession::SuspendWriteEvent()
         THROWS;
         GC_TRIGGERS;
         MODE_PREEMPTIVE;
+        // Need to disable the session before calling this method
+        PRECONDITION(HasOnlyRundownProviders());
     }
     CONTRACTL_END;
 
-    // Force all in-progress writes to either finish or cancel
-    // This is required to ensure we can safely flush and delete the buffers
-    m_pBufferManager->SuspendWriteEvent(GetIndex());
+    {
+        SpinLockHolder holder(EventPipeThread::GetGlobalThreadLock());
+
+        EventPipeThreadIterator eventPipeThreads = EventPipeThread::GetThreads();
+        EventPipeThread *pThread = nullptr;
+        while (eventPipeThreads.Next(&pThread))
+        {
+            // Wait for the thread to finish any writes to this session
+            YIELD_WHILE(pThread->GetSessionWriteInProgress() == GetIndex());
+
+            // Since we've already disabled the session, the thread won't call back in to this
+            // session once its done with the current write
```

Comment:
I think the logic needs a small adjustment. Imagine:

1. Event writer thread: executed up to but not including [this line](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L800). 
2. Disable() thread preempts the writer, and executes down to line 586 because no threads report a write in progress.
3. Event writer thread peempts Disable() thread: Executes [Line 802](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L802) loading a non-null session pointer
4. Disable() thread preempts writer: executes the remainder of EventPipe::Disable, deletes session
5. Event writer thread preempts Disable() thread: [Calls EventPipeSession::WriteEventBuffered()](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L807) with deleted session pointer, AVs 

The issue appears to be [storing a null pointer into the session array](https://github.com/dotnet/runtime/blob/5f87367da4b55f86410e1e820bd6a4b6439c2a16/src/coreclr/src/vm/eventpipe.cpp#L552) occurs too late. The null store must happen prior to this scan.

After that I think the algorithm is sound:
Writer thread does:
1. [Write InProgress bit = true](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L800)
2. [Read session pointer](https://github.com/dotnet/runtime/blob/7a8da1a706bd011ec80e7cab28dd0ad38cf54cc3/src/coreclr/src/vm/eventpipe.cpp#L802)

Disable thread does:
1. Write session pointer = null
2. Read InProgress bit (579 above)

In any possible interleaving at least one of the writes must occur before the read on the other thread. If Disable() observes the InProgress bit it waits at least long enough for the write to complete (possibly longer, potentially even allowing multiple new event writes to occur, but the same logic can be applied recursively) so we don't delete the pointer while it is in use. If WriteEventInternal() observes the null session pointer then it bails out without invoking WriteEventBuffered()

Reply:
Thanks Noah. I just pushed changes to move the NULL store to before the call to pSession->SuspendWriteEvent